src/Microsoft.OData.Core/UriParser/SemanticAst/ODataPath.cs

```diff
@@ -94,9 +107,18 @@ internal IList<ODataPathSegment> Segments
         /// Get the segments enumerator
         /// </summary>
         /// <returns>The segments enumerator</returns>
-        public IEnumerator<ODataPathSegment> GetEnumerator()
+        public Enumerator GetEnumerator()
         {
-            return this.segments.GetEnumerator();
+            return new Enumerator(this.segments);
+        }
+
+        /// <summary>
+        /// Get the segments enumerator
+        /// </summary>
+        /// <returns>The segments enumerator</returns>
+        IEnumerator<ODataPathSegment> IEnumerable<ODataPathSegment>.GetEnumerator()
+        {
+            return this.GetEnumerator();

```

Comment:
I think this implementation should remain as `return this.segments.GetEnumerator()`. We will have to do boxing either way, but calling `this.GetEnumerator()` will add an extra level of indirection for each call to the `IEnumerator`

Reply:
I'm happy to go either way - I opted for this because that way the same enumeration logic is used for all `GetEnumerator` variations, but since it's a simple enumerator that really isn't a strong argument.

Reply:
I'm reverting my enumeration change - the indexer gets me what I need perf wise